flowmachine/flowmachine/features/subscriber/topup_balance.py

```diff
@@ -0,0 +1,234 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# -*- coding: utf-8 -*-
+"""
+Class for calculating top-up statistics.
+"""
+
+import warnings
+
+from ..utilities.sets import EventsTablesUnion
+from .metaclasses import SubscriberFeature
+
+valid_stats = {"count", "sum", "avg", "max", "min", "median", "stddev", "variance"}
+
+
+class TopUpBalance(SubscriberFeature):
+    """
+    This class calculates statistics associated with top-up balances.
+
+    Top-up balance is a stock variable. As such, here we calculate the weighted
+    balance, weighted by the number of seconds a subscriber held that balance.
+    Given that we only learn about changes in balance when a top-up event
+    occurs, this average will be biased upwards. For instance, if a subscriber
```

Comment:
Think it is worth clarifying _why_ this is the case, because at first glance it seems like we should be able to work out their balance given we have info on their calls, texts etc. Obviously reason is because we don't know what those cost them, but worth mentioning?